### DISCUSSION TITLE
Join with virtual entity

### DISCUSSION THREAD

**Discussion by dirslashls** [mikro-orm/mikro-orm#4823] [General]
It seems like it is not possible to join to virtual entities. Is that correct? The motivation for this requirement is that if virtual entities can be thought of as database views, then it is possible to join tables and views. In my specific case, I have a JSON property in a table and then I would like to use JSON_TABLE() function to do a lateral join and expand one row in the table containing JSON into multiple rows. By abstracting this logic into a virtual entity, I was hoping the rest of joins to other tables could be done using the QB API.

**Comment by B4nan** [comment]
Yes, virtual entity has no identity, you cant join it directly. What you can do nowadays is to define a formula property as a relation - the expression will give you the FK, search the issues for virtual relations with formulas.

**Reply by B4nan** [reply]
See https://github.com/mikro-orm/mikro-orm/issues/4759

**Comment by dirslashls** [comment]
Thanks for the pointers @B4nan . I reviewed and tried multiple ways but am still struggling. I am not sure if currently what I want to achieve is possible or not.

I have a `user` entity and a `workbook` entity and then there is a many-to-many relationship between them stored in a `recent_data` entity. This table has `user` as a column and `recent` as a JsonType property which is an array of `workbook` ids (there is also an `entity` column which allows capturing relationship of various entities to the user with `workbook` as one of them but I think we can ignore this generalization for this discussion).

```sql
CREATE TABLE `recent_data` (
  `user_id` varchar(32) NOT NULL,
  `entity` varchar(32) NOT NULL,
  `recent` json NOT NULL,
  PRIMARY KEY (`user_id`,`entity`)
)
```

What I need is the ability to query the list of recent workbooks for a given user. The SQL would be something like

```sql
select wb.id,wb.name,ids.accessed_date
from recent_data rd
     ,JSON_TABLE(rd.recent,'$[*]' COLUMNS(id varchar(32) CHARACTER SET utf8mb4 path '$.id',accessed_date datetime path '$.ts')) ids
     ,workbook wb
where wb.id = ids.id
  and rd.entity = 'workbook'
  and rd.user_id = ?
```
and I could use em.getConnection() to directly execute the query

```js
em.getConnection().execute(recentWbSQL,[User.getUserId()],'all');
```

I am trying to make use of the provided ORM constructs to achieve this rather than doing it all using raw sql.

**Reply by B4nan** [reply]
Don't expect any `json_table` magic, if you need that, you need to use raw queries. The ORM will never produce such a query, it will never make a multi-table select query without joining in the first place.

What I suggested is a way to map expression (returning a FK) to a relation property, as that is (the only) way to have a virtual relation.
